compiler/ast/src/main/kotlin/motif/ast/compiler/CompilerType.kt

```diff
@@ -81,9 +81,12 @@ class CompilerType(
             return type
         }
 
-        val superType = MoreTypes.nonObjectSuperclass(env.typeUtils, env.elementUtils, type).orNull() ?: return null
+        env.typeUtils.directSupertypes(type).forEach { superType ->
```

Comment:
This could be simplified and made safer to 

```kotlin
return env.typeUtils.directSupertypes(type)
    .asSequence()
    .filters instance<DeclaredType>()
    .mapNotNull { getMatchingType(baseType, it) }
    .firstOrNull()
```

Reply:
Also - is this necessary? On mobile but there is definitely some API on Types like `isAssignable(Type from, Type to)`

Reply:
We actually don't want to filter out DeclaredTypes here. We should fail fast instead, so going to keep the cast until we have a holistic error type strategy. Will make the sequence change though.

Reply:
You could move that cast into the mapNotNull body then and remove the filter

Reply:
Updated